### PR TITLE
Allow stealing of fast tasks in some situations

### PR DESCRIPTION
**Pull request by mrocklin**
Previously we avoided the stealing of fast tasks in some situations.  It
was generally considered a bad idea to spend non-trivial time in order
to move around a 1ms task.  However, sometimes it's not this task that
matters, but the tasks that it unblocks.  Being strict about not
stealing may not always be ideal.

This commit relaxes stealing of fast tasks in three ways:

1.  It sets a minimum duration of tasks to 5ms,
    this being something like the overhead in a real system

2.  It changes the network latency variable in stealing.py from 100ms to 10ms

3.  It no longer completely rules out very fast tasks from stealing, but
    instead lets them compete based on their compute/transfer ratio
    (which should ordinarily be terrible)

    In cases where transfer times are trivial this becomes doable again.

Tests don't pass yet, this is up for comments

Fixes https://github.com/dask/distributed/issues/4471

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @fjetter @gjoseph92

File: distributed/stealing.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
 # submission which may include code serialization. Therefore, be very
 # conservative in the latency estimation to suppress too aggressive stealing
 # of small tasks
-LATENCY = 0.1
```
Comment (fjetter) [inline distributed/stealing.py:26]: I don't think this is realistic, see https://github.com/dask/distributed/pull/5390

File: distributed/stealing.py [hunk 2]
```diff
@@ -199,8 +199,7 @@ def steal_time_ratio(self, ts):
 
         ws = ts.processing_on
         compute_time = ws.processing[ts]
-        if compute_time < 0.005:  # 5ms, just give up
-            return None, None
+        compute_time = max(compute_time, 0.010)
```
Comment (fjetter) [inline distributed/stealing.py:202]: I've been down this road before, see https://github.com/dask/distributed/pull/4920

I also noticed the way we measure bandwidths is very skewed which impacts this significantly, see https://github.com/dask/distributed/issues/4962#issuecomment-867821099 Not sure if this is still relevant but it was big back then

I still think https://github.com/dask/distributed/pull/4920 is worth getting in but I had to drop this back then because deadlocks popped up.

**Review by gjoseph92 — COMMENTED** [review]
Stealing small tasks (like the splits of a map_overlap or a shuffle) might be a good idea in some special cases (like scale-up), but it's a bad idea in others. I'm wary of just changing some magic numbers here to make this case work.

Isn't this change a bit like a different way of implementing https://github.com/dask/distributed/pull/4925? There was a good reason not to merge that https://github.com/dask/distributed/pull/4925#issuecomment-883716281; I imagine it might apply to this too.

I think the conversation in https://github.com/dask/distributed/issues/4471#issuecomment-861483765 was going in the right direction in identifying flaws in the stealing and `decide_worker` heuristics during scale-up.

To me, the reason to steal a map_overlap split onto a new worker doesn't have to do with the task's duration, it has to do with the duration of the dependencies it opens up, and with the imbalance in occupancy between workers. So it would feel odd if those factors didn't come into play in how we change the decision.

File: distributed/scheduler.py [hunk 3]
```diff
@@ -3441,7 +3441,7 @@ def get_task_duration(self, ts: TaskState) -> double:
         """
         duration: double = ts._prefix._duration_average
         if duration >= 0:
-            return duration
+            return max(0.005, duration)
```
[the remark this answers is not in the document]
Comment (mrocklin) [reply distributed/scheduler.py:3444]: That's true.  It's not unreasonable though.  getitem tasks have durations like 62us, which I think is not reflective of general overhead.  I think that it probably makes sense for us to set a minimum to our compute times more broadly.

**Comment by mrocklin** [comment]
> Stealing small tasks (like the splits of a map_overlap or a shuffle) might be a good idea in some special cases (like scale-up), but it's a bad idea in others. I'm wary of just changing some magic numbers here to make this case work.

Yeah, sorry, I probably need to learn to use the draft PR feature more often.  Often I'll throw up PRs just to get thoughts on them.  This is one.  Please don't interpret this as "I plan to merge this right now".  Indeed, based on @fjetter 's comments I'm more inclined to go play with https://github.com/dask/distributed/pull/4920

Stealing fast tasks is often a bad idea, yes, but we make it artificially bad.  We might want to relax some of those constraints.

**Comment by mrocklin** [comment]
> Isn't this change a bit like a different way of implementing https://github.com/dask/distributed/pull/4925? There was a good reason not to merge that https://github.com/dask/distributed/pull/4925#issuecomment-883716281; I imagine it might apply to this too.

Thanks for pointing me back there.  I'll take a look.

**Comment by fjetter** [comment]
I want to emphasise that whatever we do is highly sensitive to the bandwidth measurements
See  https://github.com/dask/distributed/issues/5324 and particularly https://github.com/dask/distributed/issues/5324#issuecomment-920712272 
which is still there https://github.com/dask/distributed/blob/41ecbca2f94f029ebcbcaa1bbe5ae059e4858865/distributed/worker.py#L2946

also
https://github.com/dask/distributed/issues/4962#issuecomment-867821099

TLDR

- I believe we should reduce the default bandwidth and keep default latencies high. overestimating costs (i.e. underestimating bandwidths, overestimating latencies) is safe and we can learn and improve while the cluster is running without making any bad decisions
- The way the scheduler measures bandwidths is slow, biased and depends on whether or not workers even participate in worker<->worker communication. E.g. if half of the workers would just be idling, they would suggest to the scheduler a very wrong bandwidth since they contribute equally to workers under heavy load. 
- Particularly latencies which dominate small transfers are very strongly underestimated by not updating any bandwidths for small transfers
- Maybe we should start also measuring latencies

I never dared to modify this without solid measurements.
